pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -0,0 +1,284 @@
+    `Jones <https://arxiv.org/abs/2011.02991>`__ to compute the metric tensor.
+
+    A mixture of a main forward pass and intermediate partial backwards passes is
+    used to evaluate the metric tensor in O(num_params^2) operations, using 4 state
+    vectors.
+
+    .. note::
+        The adjoint metric tensor method has the following restrictions:
+
+        * As it requires knowledge of the statevector, only statevector simulator
+          devices can be used.
+
+        * We assume the circuit to be composed of unitary gates only and rely
+          on the ``generator`` property of the gates to be implemented.
+          Note also that this makes the metric tensor strictly real-valued.
+
+    Args:
+        circuit (.QuantumTape or .QNode): Circuit to compute the metric tensor of
+        device (.Device): Device to use for the adjoint method
+        hybrid (bool): Whether to take classical preprocessing into account. Ignored if
+            ``circuit`` is a tape.
+
+    Returns:
+        array: the metric tensor of the tape with respect to its trainable parameters.
+        Dimensions are ``(len(trainable_params), len(trainable_params))``.
+
+    """
+    if isinstance(circuit, qml.tape.QuantumTape):
+        return _adjoint_metric_tensor_tape(circuit, device)
+    if isinstance(circuit, (qml.QNode, qml.ExpvalCost)):
+        return _adjoint_metric_tensor_qnode(circuit, device, hybrid)
+
+    raise qml.QuantumFunctionError("The passed object is not a QuantumTape or QNode.")
+
+
+def _adjoint_metric_tensor_tape(tape, device):
+    """Computes the metric tensor of a tape using the adjoint method and a given device."""
+    # pylint: disable=protected-access
+    if device.shots is not None:
+        raise ValueError(
+            "The adjoint method for the metric tensor is only implemented for shots=None"
+        )
+    # original_parameters = _prepare_tape_params(tape, device, interface)
+    tape = qml.transforms.expand_trainable_multipar(tape)
+
+    # Divide all operations of a tape into trainable operations and blocks
+    # of untrainable operations after each trainable one.
+    trainable_operations, group_after_trainable_op = _group_operations(tape)
+
+    # generate and extract initial state
+    psi = device._create_basis_state(0)  # pylint: disable=protected-access
+    dim = 2 ** device.num_wires
+
+    num_params = len(tape.trainable_params)
+    # initialize metric tensor components (which all will be real-valued)
+    _like_real = qml.math.real(psi[0])
+    L = qml.math.convert_like(qml.math.zeros((num_params, num_params)), _like_real)
+    T = qml.math.convert_like(qml.math.zeros((num_params,)), _like_real)
+
+    psi = _apply_any_operation(psi, group_after_trainable_op[-1], device)
+
+    for j, outer_op in enumerate(trainable_operations):
+        # sub_L = []
+        generator_1, prefactor_1 = _get_generator(outer_op)
+
+        # the state vector phi is missing a factor of 1j * prefactor_1
```

Comment:
how come?

Reply:
Because state vector operations are expensive and we can just "bookkeep" this factor in the commentary of the code and multiply the inner products we get with those factors, instead of exponentially large vectors :D